command(): read larger chunks at a time, keep the bus active

liamstask opened the pull request:
the response header tells us the payload length, so attempt to read it all in one chunk to avoid issuing several bulk read requests.

also, renames PtpTransaction to PtpContainerInfo since a) PtpTransaction was a misnomer and b) container terminology is unified with existing `PtpContainerType`

liamstask committed:
command(): read larger chunks at a time, keep the bus active

the response header tells us the payload length, so attempt to read it all in one chunk to avoid issuing several bulk read requests.

also, renames PtpTransaction to PtpContainerInfo since a) PtpTransaction was a misnomer and b) container terminology is unified with existing `PtpContainerType`

liamstask commented on src/lib.rs:
```diff
@@ -942,56 +928,70 @@ impl<'a> PtpCamera<'a> {
+            trace!("  bulk rx {}, ({}/{})", n, buf.len(), buf.capacity());
         }
+
+        Ok((cinfo, buf.split_off(PTP_CONTAINER_INFO_SIZE)))
```
eh, hold on - i see `split_off()` allocates and copies, will improve.

kevinmehall replied:
Yeah, you could cut off the header before reading the rest of the payload, and it would also clean up the math around `additional`:

``` rust
let mut payload = Vec::with_capacity(cinfo.payload_len);
payload.extend_from_slice(&buf[PTP_CONTAINER_INFO_SIZE..]);
```

liamstask replied:
yep, though i was considering `buf.drain(..PTP_CONTAINER_INFO_SIZE)` after parsing the header info, which is O(1) (I think) while `extend_from_slice()` is O(n) for the length of the slice it's extending from.

kevinmehall replied:
No, drain still shifts up the part of the buffer after the drained range. It's only an optimization over calling `remove` repeatedly, which shifts for every element removed.

Cutting elements off the beginning of a Vec inherently requires a copy, because the first element and the start of allocation are the same pointer.

liamstask replied:
gotcha, so the entire `for i in 0..other.len()` loop within `extend_from_slice()` is expected to be optimized to a `memcpy()/memmove()`?

I was put off by the docs that say `Iterates over the slice 'other', clones each element, and then appends it to this 'Vec'.`

kevinmehall replied:
Yes, LLVM is pretty good at that.

kevinmehall commented on src/lib.rs:
```diff
@@ -791,42 +791,36 @@ impl PtpPropInfo {
+    pub kind: PtpContainerType,
+    pub tid: u32,               // transaction ID that this container belongs to
+    pub code: u16,              // StandardCommandCode or ResponseCode, depending on 'kind'
+    pub payload_len: usize,     // payload len in bytes, usually relevant for data phases
```
Make these doc comments (`///`, on the line before the item)

kevinmehall replied:
though the struct isn't public, so they wouldn't show in rustdoc

liamstask replied:
does it make sense for the fields to be `pub` at all in this case, then?

kevinmehall replied:
not really

kevinmehall commented on src/lib.rs:
```diff
@@ -942,56 +928,70 @@ impl<'a> PtpCamera<'a> {
+            slice::from_raw_parts_mut(buf.as_mut_ptr(), buf.capacity())
+        };
+        let n = try!(self.handle.read_bulk(self.ep_in, bufslice, timeout));
+        unsafe { buf.set_len(n); }
```
It's considered good practice (but doesn't compile differently) to restore invariants before leaving `unsafe{}`, so I would put these three lines in the same block.

liamstask replied:
will do

kevinmehall commented on src/lib.rs:
```diff
@@ -942,56 +928,70 @@ impl<'a> PtpCamera<'a> {
+        while cinfo.payload_len > buf.len() - PTP_CONTAINER_INFO_SIZE {
+            // allocate one extra to deal w/zero length packets appropriately
+            let additional = cinfo.payload_len + PTP_CONTAINER_INFO_SIZE - buf.len() + 1;
+            buf.reserve(additional);
```
This wants to be outside of the loop. (or replaced; see comment below)

kevinmehall commented on src/lib.rs:
```diff
@@ -942,56 +928,70 @@ impl<'a> PtpCamera<'a> {
+        trace!("container {:?}, bulk rx {}", cinfo, n);
+
+        // response didn't fit into our original buf? resize and read the rest
+        // should ideally require a single read, but loop to handle short reads.
```
Does the PTP spec allow short packets in the middle of a payload? That's the only case I know of where libusb returns success while reading fewer bytes than requested.

liamstask replied:
looks like a short packet should only be sent to indicate the end of a command phase, so could otherwise treat it as an error. will update.

liamstask committed:
PtpContainerInfo: remove pub fields from non-pub struct, but add doc comments anyway

liamstask committed:
read_txn_phase(): only issue one additional read per response

a short read indicates end of packet, anything else should be reported as an error

also, group unsafe blocks together

liamstask committed:
command(): treat mismatched transaction IDs as an error

liamstask commented:
updated

kevinmehall commented on src/lib.rs:
```diff
@@ -967,34 +967,32 @@ impl<'a> PtpCamera<'a> {
+        trace!("container {:?}", cinfo);
+
+        // allocate one extra to avoid a separate read for trailing short packet
+        let mut payload = Vec::with_capacity(cinfo.payload_len + 1);
```
If the payload is empty, maybe return early to avoid allocating one byte.

liamstask replied:
indeed, good call

liamstask replied:
updated in 31051bb, which also avoids allocating the initial buf on the heap

kevinmehall commented on src/lib.rs:
```diff
@@ -942,56 +931,67 @@ impl<'a> PtpCamera<'a> {
+        let mut payload = Vec::with_capacity(cinfo.payload_len + 1);
+        payload.extend_from_slice(&buf[PTP_CONTAINER_INFO_SIZE..]);
+
+        // response didn't fit into our original buf? resize and read the rest
```
"resize" part is not technically true after latest changes.

liamstask replied:
edited in 31051bb

liamstask committed:
read_txn_phase(): optimize

allocate initial buf on stack, to avoid allocating for transfers with no/small payloads

early out (and avoid allocating) if response has no payload

kevinmehall commented on src/lib.rs:
```diff
@@ -942,56 +931,74 @@ impl<'a> PtpCamera<'a> {
-                
-                return Ok(msg.data);
+        unsafe {
+            let bufslice = slice::from_raw_parts_mut(&mut buf[0] as &mut u8, buf.len());
```
You can just pass `&mut buf[..]` to read_bulk and don't need `from_raw_parts_mut`. You also don't need unsafe{} because you're no longer using uninitialized memory. If you want to avoid zero-initializing buf, you can initialize it with `std::mem::uninitialized()`.

liamstask replied:
updated
